### PR TITLE
Improve dotted line for canvas

### PR DESCRIPTION
**Pull request by silamon**
Part of #4060.

### CODE REVIEW

File: addons/xterm-addon-canvas/src/TextRenderLayer.ts [hunk 1]
```diff
@@ -230,79 +230,145 @@ export class TextRenderLayer extends BaseRenderLayer {
     ctx.restore();
   }
 
+  /**
+   * Draws the foreground for a specified range of columns. Tries to batch adjacent
+   * cells of the same color together to reduce draw calls.
+   */
   private _drawForeground(firstRow: number, lastRow: number): void {
+    const cols = this._bufferService.cols;
```
Comment (Tyriar) [inline addons/xterm-addon-canvas/src/TextRenderLayer.ts:238]: Are these changes all to do with reducing draw calls? I worry it's getting a bit complicated when canvas should be used as a fallback when webgl isn't available.

File: addons/xterm-addon-canvas/src/BaseRenderLayer.ts [hunk 2]
```diff
@@ -232,15 +232,12 @@ export abstract class BaseRenderLayer implements IRenderLayer {
...
     const xLeft = x * this._scaledCellWidth;
     const yMid = (y + 1) * this._scaledCellHeight - window.devicePixelRatio - 1;
     this._ctx.moveTo(xLeft, yMid);
-    for (let xOffset = 0; xOffset < width; xOffset++) {
-      // const xLeft = x * this._scaledCellWidth;
-      const xRight = (x + width + xOffset) * this._scaledCellWidth;
-      this._ctx.lineTo(xRight, yMid);
-    }
+    const xRight = (x + width) * this._scaledCellWidth;
+    this._ctx.lineTo(xRight, yMid);
```
Comment (Tyriar) [inline addons/xterm-addon-canvas/src/BaseRenderLayer.ts:240]: How does this look for wide characters? I was a little confused when working with setLineDash as it was stretching it unexpectedly when drawing across multiple cells.

Comment (silamon) [reply addons/xterm-addon-canvas/src/BaseRenderLayer.ts:240]: before 
image

after
image

File: addons/xterm-addon-canvas/src/TextRenderLayer.ts (hunk 1 above)
Comment (silamon) [reply addons/xterm-addon-canvas/src/TextRenderLayer.ts:238]: Any suggestion on how to improve this code or if it's worth trying to get this pull request done? All code changes are related to the initial issue and the idea of this pull request is to copy the idea of the drawBackground function just above it.

File: addons/xterm-addon-canvas/src/BaseRenderLayer.ts (hunk 2 above)
Comment (the-j0k3r) [reply addons/xterm-addon-canvas/src/BaseRenderLayer.ts:240]: 4:1m straight The lowercase g is now touching the underline, before there was a very small gap which looks better. These underline things always going to headbutt with lower case pgy and friends, not ideal when they overlap.

**Review by Tyriar — COMMENTED** [review]
Unfortunately too many things have changed in this area so there are too many conflicts. I tried to switch the `[dpr * 2, dpr]` on master which I think is the essence of this PR, but it ended up with this problem:

image

Do we need to more manually draw the pixels for dotted to ensure it looks correct and connects properly across cells at all zoom levels?

**Comment by silamon** [comment]
> Unfortunately too many things have changed in this area so there are too many conflicts. I tried to switch the `[dpr * 2, dpr]` on master which I think is the essence of this PR, but it ended up with this problem:
> 
> image
> 
> Do we need to more manually draw the pixels for dotted to ensure it looks correct and connects properly across cells at all zoom levels?

That problem was also solved in this PR by drawing one line from "a" to "h" instead of for every character.

**Comment by jerch** [comment]
On a sidenote - I had a similar issue in the image addon for pixel perfect alignment. My solution there is quite simple - the addon overdraws by flooring to the left and ceiling to the right side, which avoid striping artifacts. I think a similar technique can be used here, where the on/off state  of dots only depend on total pixel progression on the viewport?

**Pull request closed without merging**